src/ECL/ECL.hs

```diff
@@ -26,11 +30,26 @@ type Binders lit = [Binder lit]
 data Binder lit
   = Var (Maybe Name)
   | Lit lit
-  | Tagged Name (Binder lit)
+  | Tagged Name [Binder lit]
   | Product (Binders lit)
   | Record [(Name, Binder lit)]
   deriving (Show, Eq)
 
+-- | Represents constructors' arities.
+type Arity = Int
+
+-- | Environment
+--
+-- The language implementor should provide an environment to let
+-- the checker lookup for constructors' names (just for one type for now).
+newtype Environment = Environment { envInfo :: Name -> [(Name, Arity)] }
+
+makeEnv :: (Name -> [(Name, Arity)]) -> Environment
+makeEnv info = Environment { envInfo = info }
+
+defaultEnv :: Environment
+defaultEnv = Environment { envInfo = (\_ -> []) }
```

Comment:
Maybe use `\_ -> Nothing` here, to indicate the name isn't defined, and add `Maybe` to `envInfo`.